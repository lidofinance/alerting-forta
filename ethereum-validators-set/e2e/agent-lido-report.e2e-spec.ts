import { configureContainer, Finding } from "forta-agent";
import { AwilixContainer, asFunction } from "awilix";
import {
  provideAgentPath,
  provideRunBlock,
  provideRunTransaction,
  removeTimestamp,
} from "./utils";

const TEST_TIMEOUT = 60_000; // ms

describe("agent-lido-report e2e tests", () => {
  let runBlock: (blockHashOrNumber: string | number) => Promise<Finding[]>;
  let runTransaction: (txHash: string) => Promise<Finding[]>;
  let logSpy: jest.SpyInstance;

  beforeAll(() => {
    logSpy = jest.spyOn(console, "log");
    logSpy.mockImplementation(() => {});
  });

  beforeEach(async () => {
    const container = configureContainer() as AwilixContainer;
    container.register({
      agentPath: asFunction(
        provideAgentPath("subagents/lido-report/agent-lido-report"),
      ),
      runTransaction: asFunction(provideRunTransaction),
      runBlock: asFunction(provideRunBlock),
    });

    // https://docs.forta.network/en/latest/cli/#invoke-commands-programmatically
    runTransaction = container.resolve("runTransaction");
    runBlock = container.resolve("runBlock");
  });

  afterAll(() => {
    jest.resetAllMocks();
  });

  it(
    "should process tx with Lido Oracle report",
    async () => {
      const findings = await runTransaction(
        "0xc786c3a7736d63da45447eb5082351d03633febe904ba5401ab59da1e89bd6ee",
      );
      expect(removeTimestamp(findings).at(0)).toMatchSnapshot();
    },
    TEST_TIMEOUT,
  );

  it(
    "should process tx with decreased Lido Beacon rewards",
    async () => {
      // todo: should be uncommented after v2 and at lest 1 decrease case since v2
      // const findings = await runTransaction(
      //   "0xc88d1cb05e76f92fc0c57fe41a13b7e8cf8b010e08a535b6f789023c001ccd4d "
      // );
      // expect(removeTimestamp(findings).at(1)).toMatchSnapshot();
    },
    TEST_TIMEOUT,
  );

  it(
    "should process tx with Lido rebase digest",
    async () => {
      const findings = await runTransaction(
        "0xbf52777e4dd583d52104be96f7da420be977faeee97cc25a89d6c81fa919056f",
      );
      expect(removeTimestamp(findings).at(0)).toMatchSnapshot();
    },
    TEST_TIMEOUT,
  );
});
