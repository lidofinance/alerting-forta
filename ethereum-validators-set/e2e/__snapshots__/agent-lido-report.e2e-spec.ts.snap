// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`agent-lido-report e2e tests should process tx with Lido rebase digest 1`] = `
Finding {
  "addresses": [],
  "alertId": "LIDO-REPORT-REBASE-DIGEST",
  "description": "*APR stats*
APR: 5.26%
Total shares: 5,644,434.78 (+38326.90) × 1e18
Total pooled ether: 6,348,106.24 (+44013.44) ETH
Time elapsed: 24 hrs 0 sec

*Rewards*
CL rewards: 623.37 (+0.25) ETH
EL rewards: 392.92 (+51.69) ETH
Total: 1,016.29 (+51.94) ETH

*Validators*
Count: 196087 (0 newly appeared)

*Withdrawn from vaults*
Withdrawal Vault: 756.29 ETH
EL Vault: 392.92 ETH

*Requests finalization*
Finalized: 42 (1,936.70 ETH)
Pending: 13 (43.68 stETH)
Share rate: 1.12467
Used buffer: 787.49 ETH

*Shares*
Burnt: 1,722.27 × 1e18",
  "labels": [],
  "metadata": {
    "apr": "5.26",
    "clRewards": "623.37",
    "clRewardsDiff": "+0.25",
    "clRewardsDiffPercent": "+0.04",
    "clValidatorsDiff": "0",
    "elRewards": "392.92",
    "elRewardsDiff": "+51.69",
    "elRewardsDiffPercent": "+15.15",
    "elWithdrawn": "392.92",
    "etherDiff": "+44,013.44",
    "finalizationBufferUsed": "787.49",
    "finalizationShareRate": "1.12467",
    "finalizedEth": "1,936.70",
    "finalizedRequestsCount": "42",
    "nonFinalizedRequestsAmount": "43.68",
    "nonFinalizedRequestsCount": "13",
    "postCLValidators": "196,087",
    "postTotalEther": "6,348,106.24",
    "postTotalShares": "5,644,434.78",
    "preTotalEther": "6,349,026.64",
    "preTotalShares": "5,646,066.68",
    "sharesBurnt": "1,722.27",
    "sharesDiff": "+38,326.90",
    "timeElapsed": "86400",
    "totalRewards": "1,016.29",
    "totalRewardsDiff": "+51.94",
    "totalRewardsDiffPercent": "+5.39",
    "wdWithdrawn": "756.29",
  },
  "name": "ℹ️ Lido Report: rebase digest",
  "protocol": "ethereum",
  "severity": 1,
  "source": {},
  "timestamp": 2023-12-31T00:00:00.000Z,
  "type": 4,
  "uniqueKey": "",
}
`;
